### PR TITLE
Change type of `$skipOnEmpty` argument in rules' constructors from `mixed` to `bool|callable|null`

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️

### CODE REVIEW

**Review by vjik — CHANGES_REQUESTED** [review]
What about another rules: `AtLeast`, `Email`, `In` ?

Need keep in constructor argument only and extract class property to class level:

```php
class {
	public function __construct(
		private mixed $skipOnEmpty = null,
	) {}
}

// ↓

class {
	private mixed $skipOnEmpty;
	public function __construct(
		bool|callable|null $skipOnEmpty = null,
	) {
		$this->skipOnEmpty = $skipOnEmpty;
	}
}
```

**Comment by arogachev** [comment]
@vjik Done. :heavy_check_mark:

File: src/Rule/Trait/SkipOnEmptyTrait.php [hunk 1]
```diff
@@ -16,7 +16,7 @@
  * ```php
  * public function __construct(
  *     // ...
- *     private mixed $skipOnEmpty = null,
+ *     bool|callable|null $skipOnEmpty = null,
  *     // ...
  * ) {
```
Comment (vjik) [inline src/Rule/Trait/SkipOnEmptyTrait.php:21]:
```suggestion
 * ) {
     $this->skipOnEmpty = $skipOnEmpty;
```

File: src/Rule/Trait/SkipOnEmptyTrait.php [hunk 2]
```diff
@@ -26,6 +26,13 @@
  */
 trait SkipOnEmptyTrait
 {
+    /**
+     * @var bool|callable|null $skipOnEmpty Whether to skip this rule if the validated value is empty / not passed.
+     *  See {@see SkipOnEmptyInterface}.
+     * @psalm-var SkipOnEmptyValue $skipOnEmpty
+     */
+    private mixed $skipOnEmpty;
```
Comment (vjik) [inline src/Rule/Trait/SkipOnEmptyTrait.php:34]:
```suggestion
    private mixed $skipOnEmpty = null;
```

File: tests/Rule/CompositeTest.php [hunk 3]
```diff
@@ -316,9 +322,13 @@ public function dataValidationFailed(): array
                 null,
                 [
                     new class () extends Composite {
-                        public function __construct()
-                        {
-                            $this->rules = [new Required()];
```
Comment (vjik) [inline tests/Rule/CompositeTest.php:321]: Keep as is. Composite rule must be easy for extend.

File: tests/Rule/CompositeTest.php [hunk 4]
```diff
@@ -219,8 +220,13 @@ public function dataValidationPassed(): array
                 20,
                 [
                     new class () extends Composite {
-                        public function __construct()
-                        {
```
Comment (vjik) [inline tests/Rule/CompositeTest.php:223]: Keep as is. Composite rule must be easy for extend.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/Rule/Trait/SkipOnEmptyTrait.php (hunk 2 above)
Comment (vjik) [inline src/Rule/Trait/SkipOnEmptyTrait.php:34]: Add info to upgrade.md for users that created own rules with this trait usage.

File: UPGRADE.md [hunk 5]
```diff
@@ -30,3 +30,36 @@ for both A and B.
...
+      // ...
+  }
+  ```
+  
+  Change to:
+
+  ```php
+  public function __construct(
+      // ...
+      private mixed $skipOnEmpty = null,
```
Comment (vjik) [inline UPGRADE.md:59]:
```suggestion
      bool|callable|null $skipOnEmpty = null,
```

**Review by vjik — APPROVED** [review]

File: UPGRADE.md [hunk 6]
```diff
@@ -30,3 +30,34 @@ for both A and B.
...
+  
+  Change to:
+
+  ```php
+  public function __construct(
+      // ...
+      bool|callable|null $skipOnEmpty = null,
+      // ...
+  ) {
+      $this->skipOnEmpty = $this->skipOnEmpty;
```
Comment (vjik) [inline UPGRADE.md:61]:
```suggestion
      $this->skipOnEmpty = $skipOnEmpty;
```

**Pull request merged**